extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js

```diff
@@ -313,9 +341,66 @@ export class AmpAdNetworkAdsenseImpl extends AmpA4A {
   }
 
   /** @override */
+  buildCallback() {
+    super.buildCallback();
+
+    this.autoFormat_ =
+        this.element.getAttribute('data-auto-format') || '';
+
+    if (this.isResponsive_()) {
+      // Attempt to resize to the correct height. The width should already be
+      // 100vw, but is fixed here so that future resizes of the viewport don't
+      // affect it.
+      const viewportSize = this.getViewport().getSize();
+      return this.attemptChangeSize(
+          AmpAdNetworkAdsenseImpl.getResponsiveHeightForContext_(
+              viewportSize),
+          viewportSize.width);
+    }
+  }
+
+  /** @override */
+  onLayoutMeasure() {
+    super.onLayoutMeasure();
+
+    if (this.isResponsive_() && !this.responsiveAligned_) {
+      this.responsiveAligned_ = true;
+
+      const layoutBox = this.getLayoutBox();
+
+      // Nudge into the correct horizontal position by changing side margin.
+      this.getVsync().measure(() => {
+        // TODO(charliereams): Is this the right way to get direction?
+        const parentDirection =
+            computedStyle(this.win, this.element)['direction'];
+        if (parentDirection == 'rtl') {
+          setStyle(this.element, 'marginRight', layoutBox.left, 'px');
```

Comment:
I think there's some confusion about vsync phases. There are two, measure and mutate, and you can't mix them together without suffering style recalcs. So we definitely need to do these two separately:

```js
this.getVsync().run({
  measure: (state) => {
    state.direction = computedStyle(this.win, this.element)['direction'];
  },
  mutate: (state) => {
    if (state.direction = 'rtl') {
      //...
    } else {
      //...
    }
  }
}, { direction: '' });